zyte_api/_async.py

```diff
@@ -0,0 +1,152 @@
+        self.retrying = retrying or zyte_api_retrying
+        self.user_agent = user_agent or USER_AGENT
+
+    async def get(
+        self,
+        query: dict,
+        *,
+        endpoint: str = 'extract',
+        session=None,
+        handle_retries=True,
+        retrying: Optional[AsyncRetrying] = None,
+    ):
+        retrying = retrying or self.retrying
+        post = _post_func(session)
+        auth = aiohttp.BasicAuth(self.api_key)
+        headers = {
+            'User-Agent': self.user_agent,
+            'Accept-Encoding': 'br'
+        }
+
+        response_stats = []
+        start_global = time.perf_counter()
+
+        async def request():
+            stats = ResponseStats.create(start_global)
+            self.agg_stats.n_attempts += 1
+
+            post_kwargs = dict(
+                url=self.api_url + endpoint,
+                json=_process_query(query),
+                auth=auth,
+                headers=headers,
+            )
+
+            try:
+                async with post(**post_kwargs) as resp:
+                    stats.record_connected(resp.status, self.agg_stats)
+                    if resp.status >= 400:
+                        content = await resp.read()
+                        resp.release()
+                        stats.record_read()
+                        stats.record_request_error(content, self.agg_stats)
+
+                        raise RequestError(
+                            request_info=resp.request_info,
+                            history=resp.history,
+                            status=resp.status,
+                            message=resp.reason,
+                            headers=resp.headers,
+                            response_content=content,
+                        )
+
+                    response = await resp.json()
+                    stats.record_read(self.agg_stats)
+                    return response
+            except Exception as e:
+                if not isinstance(e, RequestError):
+                    self.agg_stats.n_errors += 1
+                    stats.record_exception(e, agg_stats=self.agg_stats)
+                raise
+            finally:
+                response_stats.append(stats)
+
+        if handle_retries:
+            request = retrying.wraps(request)
+
+        try:
+            # Try to make a request
+            result = await request()
+            self.agg_stats.n_success += 1
+        except Exception:
+            self.agg_stats.n_fatal_errors += 1
+            raise
+
+        return result
+
+    async def iter(
+        self,
+        queries: List[dict],
+        *,
+        endpoint: str = 'extract',
+        session: Optional[aiohttp.ClientSession] = None,
+        handle_retries=True,
+        retrying: Optional[AsyncRetrying] = None,
+    ) -> AsyncGenerator[Union[dict, Exception], None]:
+        """ Send multiple requests to Zyte API in parallel.
+        Return an `asyncio.as_completed` iterator.
```

Comment:
Should we document the behavior for the exceptions?

It also looks like the example in the docs handles exceptions in a different way (via try/except), while the new code yields the exceptions. What's the advantage of yielding the exceptions here?

Reply:
I think no (see the example in the current docs). I think the the difference with the sync version is that we return an iterable of awaitables, and the exception happens when user awaits the awaitable - you can catch it in the code. In the sync version there is no awaitable to await, so it stops the iteration.